### PR TITLE
Restore snapshot download & op-mainnet snapshot

### PR DESCRIPTION
**Pull request by deffrian**
Fixes Closes Resolves #
## Changes

- Op-mainnet genesis snapshot
- Now snapshot will continue downloading after restart
- Fixes op-plugin startup issue

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [ ] Yes
- [x] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs [hunk 1]
```diff
@@ -15,4 +15,10 @@ public interface ISnapshotConfig : IConfig
 
     [ConfigItem(Description = "The SHA-256 checksum of the snapshot file.", DefaultValue = "null")]
     public string? Checksum { get; set; }
+
+    [ConfigItem(Description = "Directory where snapshot will be stored.", DefaultValue = "snapshot")]
+    public string SnapshotDirectory { get; set; }
+
+    [ConfigItem(Description = "Snapshot file name.", DefaultValue = "snapshot.zip")]
+    public string SnapshotFileName { get; set; }
```
Comment (jmederosalvarado) [inline src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs:23]: let's take the SnapshotFileName from the last path segment in the snapshot url, wdyt?

Comment (deffrian) [reply src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs:23]: It should resume downloading without problems. If you use different config params

Comment (jmederosalvarado) [reply src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs:23]: yeah, but you need to change the `SnapshotFileName` if you change the `DownloadUrl`, not a big deal I guess

File: src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs [hunk 2]
```diff
@@ -15,4 +15,10 @@ public interface ISnapshotConfig : IConfig
 
     [ConfigItem(Description = "The SHA-256 checksum of the snapshot file.", DefaultValue = "null")]
     public string? Checksum { get; set; }
+
+    [ConfigItem(Description = "Directory where snapshot will be stored.", DefaultValue = "snapshot")]
+    public string SnapshotDirectory { get; set; }
+
+    [ConfigItem(Description = "Snapshot file name.", DefaultValue = "snapshot.zip")]
```
Comment (rubo) [inline src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs:22]: As this goes to the docs, we need to pay more attention to wording. Also, it depends on what the above parameter is.

```suggestion
    [ConfigItem(Description = "The name of the snapshot file.", DefaultValue = "snapshot.zip")]
```

File: src/Nethermind/Nethermind.Api/NethermindApi.cs [hunk 3]
```diff
@@ -230,7 +230,6 @@ public ISealEngine SealEngine
         public IReadOnlyList<INethermindPlugin> Plugins { get; } = new List<INethermindPlugin>();
         public IList<IPublisher> Publishers { get; } = new List<IPublisher>(); // this should be called publishers
         public CompositePruningTrigger PruningTrigger { get; } = new();
-        public ISnapProvider? SnapProvider { get; set; }
```
[the remark this answers is not in the document]
Comment (deffrian) [reply src/Nethermind/Nethermind.Api/NethermindApi.cs:233]: It's not used

File: src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs [hunk 4]
```diff
@@ -15,4 +15,10 @@ public interface ISnapshotConfig : IConfig
 
     [ConfigItem(Description = "The SHA-256 checksum of the snapshot file.", DefaultValue = "null")]
     public string? Checksum { get; set; }
+
+    [ConfigItem(Description = "Directory where snapshot will be stored.", DefaultValue = "snapshot")]
+    public string SnapshotDirectory { get; set; }
```
[the remark this answers is not in the document]
Comment (deffrian) [reply src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs:20]: both I believe, it works exactly like BaseDbPath or LogDirectory

File: src/Nethermind/Nethermind.Api/NethermindApi.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Api/NethermindApi.cs:233]: Not used. The refactor encapsulate sync code.

File: src/Nethermind/Nethermind.Optimism/OptimismPlugin.cs [hunk 5]
```diff
@@ -140,14 +138,6 @@ public Task InitSynchronization()
         _beaconPivot = new BeaconPivot(_syncConfig, _api.DbProvider.MetadataDb, _api.BlockTree, _api.LogManager);
         _beaconSync = new BeaconSync(_beaconPivot, _api.BlockTree, _syncConfig, _blockCacheService, _api.LogManager);
         _api.BetterPeerStrategy = new MergeBetterPeerStrategy(null!, _api.PoSSwitcher, _beaconPivot, _api.LogManager);
-
-        _api.SyncModeSelector = new MultiSyncModeSelector(
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Optimism/OptimismPlugin.cs:144]: I'm guessing this was the fix?

File: src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs [hunk 6]
```diff
@@ -114,8 +112,8 @@ private async Task InitDbFromSnapshot(CancellationToken cancellationToken)
         await using FileStream snapshotFileStream = new(
```
[the remark this answers is not in the document]
Comment (deffrian) [reply src/Nethermind/Nethermind.Init.Snapshot/InitDatabaseSnapshot.cs:120]: done

File: src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs [hunk 7]
```diff
@@ -15,4 +15,10 @@ public interface ISnapshotConfig : IConfig
 
     [ConfigItem(Description = "The SHA-256 checksum of the snapshot file.", DefaultValue = "null")]
     public string? Checksum { get; set; }
+
+    [ConfigItem(Description = "Directory where snapshot will be stored.", DefaultValue = "snapshot")]
```
Comment (rubo) [inline src/Nethermind/Nethermind.Init.Snapshot/ISnapshotConfig.cs:19]:
```suggestion
    [ConfigItem(Description = "The path to the directory to store the snapshot file.", DefaultValue = "snapshot")]
```

**Review by jmederosalvarado — APPROVED** [review]

**Review by asdacap — APPROVED** [review]

**Pull request merged**